src/source.ts

```diff
@@ -27,68 +28,99 @@ export default class Source extends React.Component<Props, void> {
 
   private id = this.props.id;
 
-  public componentWillMount() {
+  private onSourceData = (e: any) => {
     const { map } = this.context;
-    const { geoJsonSource, tileJsonSource, onSourceAdded } = this.props;
-
-    if (!map.getSource(this.id) && (geoJsonSource || tileJsonSource)) {
-      map.addSource(this.id, (geoJsonSource || tileJsonSource) as any);
-      map.on('load', this.onData);
-
-      if (onSourceAdded) {
-        onSourceAdded(map.getSource(this.id) as GeoJSONSource | TilesJson);
+    const { id, props } = this;
+    if (e.isSourceLoaded && e.sourceId === this.props.id) {
+      // Refresh data (sometimes unloade without apparent reason)
+      const source = map.getSource(id) as GeoJSONSource | TilesJson;
+      if (source && props.geoJsonSource && props.geoJsonSource.data) {
+        const geoJSONSource = source as GeoJSONSource;
+        geoJSONSource.setData(props.geoJsonSource.data);
```

Comment:
you don't need to do `const geoJSONSource = source as GeoJSONSource;` you can just do `(source as GeoJSONSource).setData`